Track refactoring

florian-kuebler opened the pull request:
### Adjust Positioning System of Tracks - Make it Self-contained.

CaptureViewElements have position and size information for their
appearance in the world coordinate system. This was only partially
used for tracks. The track's header/tab was not included by the
position/size of the track itself. Instead, the layout's gap size
between tracks contained the size of the header. The track relied
on this implicit space and drawn the header out of its scope.

This change adjusted the track's size. It is now the complete size
of a track.

Further, -- since there were bugs and it was unused -- this change
removes the handling for pinned tracks. The were drawn differently,
and remain untested.

Follow-up:
1. Simplify the Accessibility Interface (for tracks).
2. The nice thing about the state before this change, was that
   drawing the track's content did not required any knowledge about
   the header (and other margins). However, the clean solution would
   be to split the Track into a TrackHeader/Tab, (possibly multiple)
   TrackContentPane and TrackFooter.

Test: Take a capture including all types of tracks.


-----

### Introduce AccessibleCaptureViewElement as base class for accessibility.

With CaptureViewElement's following the same positioning (with
their pos/size set correctly) as well as adding knowledge about
their parent, this change extracts some computations for accessibility
into a base class.

Follow-ups:
1. Remove the need for virtual controls, by splitting Track into
   multiple CaptureViewElements.
2. Let the AccessibleCaptureViewElement take more parameters,
   instead of overriding for specialisations.

Test: Use AccessibilityInsights.

_____
### AccessibilityInterface now uses absolute positions for the rect.

Postions of CaptureViewElements as well as in AccessibilityAdapter::rect()
are absolute. Prior this change, we converted CaptureViewElement's position
to a relative one in order to convert it back to an absolute one.
This makes the AccessibilityInterfaces unnecessarily complicated.

This change switch also to absolute positions in the accessibility.
Further, it introduces FakeTimerPane and FakeTrackTab as capture view
elements. Now all position calculations for accessibility are done
by AccessibleCaptureViewElement.

Follow up: The fake classes should be actual children of a Track.
  This requires the actual drawing to be also moved to those classes.

Test: Take a capture and inspect with the AccessibilityInsights.

florian-kuebler commented:
Best to review each commit separately.

florian-kuebler commented:
Added @vickyliu-go4it FYI

ronaldfw commented:
Does this have to be a single PR? I haven't looked at detail in the code, maybe there is a dependency I'm missing that requires that these are done as one PR. But it would make reviewing easier if this were split into two PRs.

florian-kuebler commented:
> Does this have to be a single PR? I haven't looked at detail in the code, maybe there is a dependency I'm missing that requires that these are done as one PR. But it would make reviewing easier if this were split into two PRs.

I thought about this. The issue is, that each commit would break the E2E tests if taking alone. This could be fixed, by adjusting the Accessebility in the first commit. But we would then end up in a very temporal change that would be nearly as large as the second commit.

ronaldfw commented:
> I thought about this. The issue is, that each commit would break the E2E tests if taking alone. This could be fixed, by adjusting the Accessebility in the first commit. But we would then end up in a very temporal change that would be nearly as large as the second commit.

Ok,  got it. Then it's probably better to do this as a single PR.

karupayun reviewed (approved):
LGTM! Thanks for addressing this issue :)

florian-kuebler commented on src/OrbitGl/Track.cpp:
```diff
@@ -135,11 +123,11 @@ void Track::Draw(GlCanvas* canvas, PickingMode picking_mode, float z_offset) {
-    Vec2 bottom_right(tab_x0 + label_width, y0 + top_margin);
-    Vec2 top_right(tab_x0 + label_width, y0 + label_height);
-    Vec2 top_left(tab_x0, y0 + label_height);
+    Vec2 bottom_right(tab_x0 + label_width, y0 - label_height + top_margin);
```
Done, PTAL.

reichlfl commented on src/OrbitGl/FrameTrack.cpp:
```diff
@@ -79,14 +79,14 @@ FrameTrack::FrameTrack(CaptureViewElement* parent, TimeGraph* time_graph, TimeGr
 
 float FrameTrack::GetYFromTimer(const TimerInfo& /*timer_info*/) const {
-  return pos_[1] - GetMaximumBoxHeight();
+  return pos_[1] - GetHeaderHeight() - GetMaximumBoxHeight();
```
This (and the overriding `GetYFromTimer`) looks like it would be more straightforward if we renamed it to something like `GetYOffsetFromTimer`, have it return a position relative to the start of the "Timer Area", and have the caller do the required math to convert this into an absolute position, right?

It's probably not something to change in this PR though, but we should keep this in mind when moving to relative positioning.

florian-kuebler replied:
That would change completely as soon as we have the ContentPane/TimerPane's. In this case, the code would look like before this change, as the pane is already positioned and all computations only need to be relative to its own position.I would keep it as it for now, as this will change anyways soon.

reichlfl commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -413,6 +413,14 @@ float ThreadTrack::GetHeaderHeight() const {
   return header_height;
 }
 
+float ThreadTrack::GetYFromDepth(uint32_t depth) const {
```
I'm not sure I understand why this override is needed now. Isn't one of the reasons `GetHeaderHeight` was introduced to avoid overriding `GetYFromDepth` for every class deriving from `TimerTrack` and provide a single template method to customize the timer positioning?

This looks like `gap_between_tracks_and_timers` should be part of the header height calculation, so the logic to check for thread bars is only required in `GetHeaderHeight`.

florian-kuebler replied:
Yeah that thing is a bit funcy, but for now, we need to keep it as it is. It is all basically because of the last space between the event bar and the timers pane.
We only want to draw it if there are timers (otherwise, "empty" tracks would have some strange space). However, in the GetHeaderHeight, we don't know about whether we have a timers pane or not.

reichlfl replied:
Ok, SGTM, let's fix it the proper way in a later PR.

reichlfl commented on src/OrbitGl/TimerTrack.cpp:
```diff
@@ -59,8 +59,7 @@ float TimerTrack::GetYFromTimer(const TimerInfo& timer_info) const {
 float TimerTrack::GetYFromDepth(uint32_t depth) const {
-  return pos_[1] - GetHeaderHeight() - layout_->GetSpaceBetweenTracksAndThread() -
-         box_height_ * static_cast<float>(depth + 1);
+  return pos_[1] - GetHeaderHeight() - box_height_ * static_cast<float>(depth + 1);
```
See my related comment above, why the change in the usage of `layout_->GetSpaceBetweenTracksAndThread()`?

florian-kuebler replied:
Here it is a bit simpler to explain: TimerTrack should not know at all about the possibility of event/thread bars above the timer pane. Only if the track itself actually has an event bar, it should handle this.

As above, it will become obsolete by the refactorings anyways.

reichlfl commented on src/OrbitGl/ThreadTrack.h:
```diff
@@ -42,6 +42,7 @@ class ThreadTrack final : public TimerTrack {
   void UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t max_tick,
                         PickingMode picking_mode, float z_offset = 0) override;
   void OnTimer(const orbit_client_protos::TimerInfo& timer_info) override;
+  [[nodiscard]] virtual float GetYFromDepth(uint32_t depth) const override;
```
... and then this would not need to be virtual, because its behavior is customized through `GetHeaderHeight`.

florian-kuebler replied:
See above

reichlfl reviewed (commented):
Thanks a lot, I think this is moving into a much clearer direction. I have a few comments - I think the only one that should be addressed is the one in `AccessibleCaptureViewElement`. The change in `GetYFromDepth` isn't ideal IMHO, but as we discussed yesterday we should handle timer positioning in another way anyways, so I'm ok with leaving it for now given that it will be temporary.

florian-kuebler commented:
I refactored the GetLocalRect to be GetRect and using absolute positions and added (partial) capture view element implementations for the TrackTab and TimerPane. Accessibility now always uses the AccessibleCaptureViewElement's positioning method.

reichlfl commented on src/OrbitQt/AccessibilityAdapter.cpp:
```diff
@@ -205,14 +205,25 @@ QAccessibleInterface* AccessibilityAdapter::childAt(int x, int y) const {
+  // Find the outer most AccessibleInterface to convert the coordinates from
+  // pixel relative to the CaptureWindow to screen positions.
+
+  const orbit_accessibility::AccessibleInterface* bridge_info = info_->AccessibleParent();
+  const QAccessibleInterface* bridge = parent();
+  while (bridge_info != nullptr && bridge_info->AccessibleParent() != nullptr) {
+    bridge_info = bridge_info->AccessibleParent();
+    bridge = bridge->parent();
+  }
```
This could be easier to read if you only trace `AccessibleParent()` upwards, and rename `bridge_info` to `root` or similar.

Once `root->AccessibleParent() == nullptr`, you should be able to retrieve `QAccessibleInterface` from `AdapterRegistry::GetOrCreateAdapter`.

At least consider naming the variables differently so it's clear you're searching for the root of our GL accessibility tree here, I think "bridge" doesn't really help to understand this.

florian-kuebler replied:
Done.

reichlfl reviewed (approved):
LGTM, thanks a lot for cleaning this up. I have some nitpicks, but I'm also fine with merging it in the current state since it's not going to be the final iteration.

florian-kuebler commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -17,188 +17,137 @@
-    return nullptr;
-  }
-  [[nodiscard]] const AccessibleInterface* AccessibleParent() const override { return parent_; }
+class FakeTrackTab : public CaptureViewElement {
```
Yes it should. See the track refactoring doc.

florian-kuebler replied:
I added a Bug for this.

florian-kuebler commented on src/OrbitGl/AccessibleThreadBar.cpp:
```diff
@@ -12,48 +12,13 @@
-  // This could be cleaned up by introducing "real" control elements for tab and content area.
-  return thread_bar_->GetParent()->GetAccessibleInterface()->AccessibleChild(1);
-}
+int orbit_gl::AccessibleThreadBar::AccessibleChildCount() const { return 0; }
```
Done. In the future, the capture view element should actually just know the childs.

florian-kuebler commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -17,188 +17,137 @@
-    return nullptr;
-  }
+    // We are looking for the track's child that is right above the timer pane.
+    CaptureViewElement* predecessor = track_tab_;
```
Done

florian-kuebler commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -219,7 +168,7 @@ AccessibilityState AccessibleTrack::AccessibleState() const {
   }
 
-  if (AccessibleLocalRect().height == 0) {
+  if (AccessibleRect().height == 0) {
```
I have the feeling the complete state handling could be simplified in the AccessibleCaptureView. I am going to omit this for now, as I don't want to get pulled into another rabbit hole.
